tensorflow/lite/experimental/micro/examples/micro_speech/apollo3evb/audio_provider.cc

```diff
@@ -0,0 +1,536 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/experimental/micro/examples/micro_speech/audio_provider.h"
+
+#include <limits>
+
+// These are headers from Ambiq's Apollo3 SDK.
+#include "am_bsp.h"         // NOLINT
+#include "am_mcu_apollo.h"  // NOLINT
+#include "am_util.h"        // NOLINT
+
+#include "tensorflow/lite/experimental/micro/examples/micro_speech/micro_features/micro_model_settings.h"
+
+namespace {
+
+// These are the raw buffers that are filled by the PDM during DMA
+constexpr int kPdmNumSlots = 1;
+constexpr int kPdmSamplesPerSlot = 256;
+constexpr int kPdmSampleBufferSize = (kPdmNumSlots * kPdmSamplesPerSlot);
+uint32_t g_ui32PDMSampleBuffer0[kPdmSampleBufferSize];
+uint32_t g_ui32PDMSampleBuffer1[kPdmSampleBufferSize];
+uint32_t g_PowerOff = 0;
+
+// Controls the double buffering between the two DMA buffers.
+int g_dma_destination_index = 0;
+// PDM Device Handle.
+static void* g_pdm_handle;
+// PDM DMA error flag.
+volatile bool g_pdm_dma_error;
+// So the interrupt can use the passed-in error handler to report issues.
+tflite::ErrorReporter* g_pdm_dma_error_reporter = nullptr;
+
+
+// Holds a longer history of audio samples in a ring buffer.
+constexpr int kAudioCaptureBufferSize = 16000;
+int16_t g_audio_capture_buffer[kAudioCaptureBufferSize] = {};
+int g_audio_capture_buffer_start = 0;
+int64_t g_total_samples_captured = 0;
+int32_t g_latest_audio_timestamp = 0;
+
+// Copy of audio samples returned to the caller.
+int16_t g_audio_output_buffer[kMaxAudioSampleSize];
+bool g_is_audio_initialized = false;
+
+
+//*****************************************************************************
+//
+// Globals
+//
+//*****************************************************************************
+#if USE_TIME_STAMP
```

Comment:
Can you explain what all these #ifdef macros are, maybe just in a comment at the top of the file? USE_TIME_STAMP, USE_MAYA, AM_BSP_NUM_LEDS, USE_DEBUG_GPIO, and any others I missed.